### PR TITLE
add more user based permission check in Memory

### PR DESCRIPTION
**Pull request by Zhangxunmt**
### Description
This PR is to add user based permission check that addresses concerns in https://github.com/opensearch-project/ml-commons/issues/1901. 
 
The permission checks are strictly applied to all interactions/conversations. 

The PR is verified using two different login users and ensure they don't have access to each others's data

More tests cases added and Rebased to latest main branch.

### Issues Resolved
[[List any issues this PR will resolve](https://github.com/opensearch-project/ml-commons/issues/1901)]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: memory/src/main/java/org/opensearch/ml/memory/action/conversation/UpdateInteractionTransportAction.java [hunk 1]
```diff
@@ -25,26 +29,47 @@
...
     @Override
     protected void doExecute(Task task, ActionRequest request, ActionListener<UpdateResponse> listener) {
+        if (!featureIsEnabled) {
+            listener
+                .onFailure(
+                    new OpenSearchException(
+                        "The experimental Conversation Memory feature is not enabled. To enable, please update the setting "
+                            + ConversationalIndexConstants.ML_COMMONS_MEMORY_FEATURE_ENABLED.getKey()
+                    )
+                );
```
Comment (HenryL27) [inline memory/src/main/java/org/opensearch/ml/memory/action/conversation/UpdateInteractionTransportAction.java:62]: Hah! I had forgotten about this part too

**Review by HenryL27 — COMMENTED** [review]
Overall looks good, thanks for doing this. 
There used to be an access test for singular get interaction - can we put it back? And maybe add some similar tests for the trace/update APIs?

**Comment by Zhangxunmt** [comment]
> Overall looks good, thanks for doing this. There used to be an access test for singular get interaction - can we put it back? And maybe add some similar tests for the trace/update APIs?

Yes those tests all be added. There might be slightly difference for the get interaction permission check due to the conversation id is removed after refactor. But overall this is a valid solution without adding too much latency (negligible from tests).

**Review by HenryL27 — APPROVED** [review]
LGTM, thanks for adding access tests

**Review by ylwu-amzn — APPROVED** [review]

**Pull request merged**